addons/link_tracker/models/link_tracker.py

```diff
@@ -183,23 +188,69 @@ def create(self, vals_list):
 
     @api.model
     def search_or_create(self, vals):
-        if 'url' not in vals:
-            raise ValueError(_('Creating a Link Tracker without URL is not possible'))
-        if vals['url'].startswith(('?', '#')):
-            raise UserError(_("%r is not a valid link, links cannot redirect to the current page.", vals['url']))
-        vals['url'] = tools.validate_url(vals['url'])
-
-        search_domain = [
-            (fname, '=', value)
-            for fname, value in vals.items()
-            if fname in ['url', 'campaign_id', 'medium_id', 'source_id']
-        ]
-        result = self.search(search_domain, limit=1)
-
-        if result:
-            return result
-
-        return self.create(vals)
+        return self._search_or_create_multi([vals])
+
+    @api.model
+    def _search_or_create_multi(self, vals_list):
```

Comment:
Thanks for doing the extra code mile :D ! Well have to say it is not crystal clear to understand ... there are indeed several use cases (create with duplicates -> try to find them back when recreating the record set to return, merge with existing one in DB, ...).

Isn't it possible to use two structures to reduce the almost-n² in _match_records_with_vals and remove loops ? Like having 'tracker_key_to_idx' (tracker tuple -> indexes in vals_list, like unique_vals_to_vals_list_index), 'tracker_key_to_id' (tracker tuple -> existing id in DB), then build new_tracker_keys, update tracker_key_to_id, and rebuild final recordset based on tracker_key_to_id ?

It looks like what you are doing (there are not 32 ways to do it :D ) but it seems we could try to make the code a bit more fluid. What do you think ?

Reply:
I made another pass as these methods and found I could simplify it a lot while keeping the same structure.
Do you think it has become more readable even though we are assigning records in place immediately instead of handling another map?

Performance wise, without building a tree of record values to find matches faster, I don't think we can do better (except perhaps in the individual record matching logic). It does take O(m log n) + O (p log p) with m= unique vals_list items, n= nb existing records and p = nb unique non existing vals (= nb records to create).

Reply:
Had troubles understanding the computation, so in the end I pushed a wip commit with a proposal that I find easier to read: fetch existing, find new unique vals to create, then rebuild the final list. Did not run the whole test suit and pushed as a separate commit so that you can have a look and say what you think.